crates/bevy_reflect/src/serde/de.rs

```diff
@@ -1425,6 +1429,37 @@ mod tests {
         assert!(expected.reflect_partial_eq(output.as_ref()).unwrap());
     }
 
+    // Regression test for https://github.com/bevyengine/bevy/issues/12462
+    #[test]
+    fn enum_should_be_correct_variant_on_reserialize() {
```

Comment:
Could you actually do this with the `MyStruct` struct? We'd then have a re-serialization test for all the major reflect kinds, including enums.